File: apple/Sources/ReactNativeWatchOS/Modifiers/Frame.swift
import SwiftUI

/// SwiftUI `.frame(...)`. Two overloads picked by which keys are present:
///   - `width` or `height`   → `.frame(width:height:alignment:)`
///   - otherwise             → `.frame(minWidth:maxWidth:…:alignment:)`
///
/// Pass `Infinity` from JS for `maxWidth` / `maxHeight` to fill available
/// space — Hermes preserves it through JSI as a Double, which CGFloat
/// converts to `.infinity` and SwiftUI accepts directly.
enum RNWFrameModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("frame") { view, params, _ in
            let alignment = parseAlignment(params.string("alignment")) ?? .center
            let w = params.cgFloat("width")
            let h = params.cgFloat("height")
            if w != nil || h != nil {
                return AnyView(view.frame(width: w, height: h, alignment: alignment))
            }
            return AnyView(view.frame(
                minWidth: params.cgFloat("minWidth"),
                maxWidth: params.cgFloat("maxWidth"),
                minHeight: params.cgFloat("minHeight"),
                maxHeight: params.cgFloat("maxHeight"),
                alignment: alignment
            ))
        }
    }

    private static func parseAlignment(_ s: String?) -> Alignment? {
        switch s {
        case "leading":        return .leading
        case "trailing":       return .trailing
        case "center":         return .center
        case "top":            return .top
        case "bottom":         return .bottom
        case "topLeading":     return .topLeading
        case "topTrailing":    return .topTrailing
        case "bottomLeading":  return .bottomLeading
        case "bottomTrailing": return .bottomTrailing
        default:               return nil
        }
    }
}

File: src/modifiers/frame.ts
import { createModifier } from './createModifier';

export type FrameAlignment =
  | 'leading'
  | 'trailing'
  | 'center'
  | 'top'
  | 'bottom'
  | 'topLeading'
  | 'topTrailing'
  | 'bottomLeading'
  | 'bottomTrailing';

export interface FrameParams {
  /// Fixed width. When set, `minWidth` / `maxWidth` are ignored.
  width?: number;
  /// Fixed height. When set, `minHeight` / `maxHeight` are ignored.
  height?: number;
  minWidth?: number;
  /// Pass `Infinity` to fill the available width (SwiftUI's `.infinity`).
  maxWidth?: number;
  minHeight?: number;
  /// Pass `Infinity` to fill the available height (SwiftUI's `.infinity`).
  maxHeight?: number;
  /// Where to position the view's content within the laid-out frame.
  /// Defaults to `center`, matching SwiftUI.
  alignment?: FrameAlignment;
}

/// SwiftUI `.frame(width:height:alignment:)` / `.frame(min/maxWidth:…)`.
/// To fill the screen, pass `{ maxWidth: Infinity, maxHeight: Infinity }`.
export function frame(params: FrameParams = {}) {
  return createModifier('frame', params);
}
